pil-analyzer/tests/condenser.rs

```diff
@@ -892,3 +892,186 @@ namespace std::prover;
+
+        namespace Main;
+            let a;
+            [a] in [b];
+            let constrs = std::prover::capture_constraints(constr || {
+                let x;
+                [x = 1, x = 2];
+                std::prover::capture_constraints(constr || {
+                    let y;
+                    [y = 1, [y] in [x]];
+                    y = 3;
+                })[1];
+                x = 3;
+            });
+            // Ignore the second constraint
+            [constrs[0], constrs[2], constrs[3]];
+            let b;
+            b = 2;
+    "#;
+    let formatted = analyze_string(input).to_string();
+    let expected = "namespace std::prover;
+    let capture_constraints: (-> ()) -> std::prelude::Constr[] = 9;
+namespace Main;
+    col witness a;
+    [Main::a] in [Main::b];
+    let constrs: std::prelude::Constr[] = std::prover::capture_constraints(constr || {
+        let x: col;
+        [x = 1, x = 2];
+        std::prover::capture_constraints(constr || {
+            let y: col;
+            [y = 1, [y] in [x]];
+            y = 3;
+        })[1];
+        x = 3;
+    });
+    col witness x;
+    col witness y;
+    Main::x = 1;
+    [Main::y] in [Main::x];
+    Main::x = 3;
+    col witness b;
+    Main::b = 2;
+";
+    assert_eq!(formatted, expected);
+}
+
+#[test]
+pub fn at_next_stage() {
+    let input = r#"
+        namespace std::prover;
+            let at_next_stage: (-> ()) -> () = 9;
+
+        namespace Main;
+            let a;
+            std::prover::at_next_stage(constr || {
+                let x;
+                std::prover::at_next_stage(constr || {
+                    let y;
+                    x = 1;
+                    y = 2;
+                });
+                let c;
+                x = a + c;
+            });
+            let b;
+    "#;
+    let formatted = analyze_string(input).to_string();
+    let expected = "namespace std::prover;
+    let at_next_stage: (-> ()) -> () = 9;
+namespace Main;
+    col witness a;
+    col witness stage(1) x;
+    col witness stage(2) y;
+    col witness stage(1) c;
+    Main::x = 1;
+    Main::y = 2;
+    Main::x = Main::a + Main::c;
+    col witness b;
+";
+    assert_eq!(formatted, expected);
+}
+
+#[test]
+pub fn at_next_stage_intermediate() {
+    let input = r#"
+        namespace std::prover;
+            let at_next_stage: (-> ()) -> () = 9;
+
+        namespace Main;
+            let a;
+            std::prover::at_next_stage(constr || {
+                let b: inter = a * a;
+                let c;
+                let d: inter = a + c;
+            });
+            let x;
+    "#;
+    let formatted = analyze_string(input).to_string();
+    let expected = "namespace std::prover;
+    let at_next_stage: (-> ()) -> () = 9;
+namespace Main;
+    col witness a;
+    col b = Main::a * Main::a;
+    col witness stage(1) c;
+    col d = Main::a + Main::c;
+    col witness x;
+";
+    assert_eq!(formatted, expected);
+}
```

Comment:
What happens with fixed columns added in a function passed to `at_next_stage`? Nothing, right? Could we add a test for that?

Reply:
I'll add a test, but it should just be added as a fixed column (who do not have stages).